tfjs-backend-wasm/src/cc/scatter_impl.cc

```diff
@@ -0,0 +1,88 @@
+/* Copyright 2020 Google LLC. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ===========================================================================*/
+
+#ifndef SCATTER_IMPL_H_
+#define SCATTER_IMPL_H_
+
+#include <cstddef>
+#include <vector>
+
+#include "tfjs-backend-wasm/src/cc/scatter_impl.h"
+
+namespace tfjs {
+namespace wasm {
+
+template <typename T>
+void scatter(const int* indices_ptr, const T* updates_ptr,
+             const size_t slice_rank, const size_t num_updates,
+             const size_t slice_size, const std::vector<size_t>& strides_ptr,
+             const size_t output_size, const T default_value,
+             const bool sum_dupe_indices, const bool update_as_scalar,
+             T* out_buf_ptr) {
+  std::fill(out_buf_ptr, out_buf_ptr + output_size, default_value);
+
+  for (size_t i = 0; i < num_updates; ++i) {
+    size_t flattened_index = 0;
+    for (size_t j = 0; j < slice_rank; ++j) {
+      flattened_index += *indices_ptr * strides_ptr[j];
+
+      indices_ptr++;
+    }
+
+    out_buf_ptr += flattened_index * slice_size;
+
+    for (size_t k = 0; k < slice_size; ++k) {
+      if (sum_dupe_indices) {
+        *out_buf_ptr += *updates_ptr;
+      } else {
+        *out_buf_ptr = *updates_ptr;
+      }
+
+      out_buf_ptr++;
+      if (!update_as_scalar) {
+        updates_ptr++;
+      }
```

Comment:
I don't fully understand the purpose of this. It looks like it's only used if `sparse_values_rank == 0` (SparseToDense.cc:35), in which case it assigns all the values of the output tensor to `*updates_ptr` (without incrementing).

Since this only happens when the rank of the sparse tensor is zero, wouldn't this only ever assign one value? In that case, I don't think it's necessary, since the loop will only run once.

On the other hand, it's [implemented the same way in the CPU backend](https://github.com/kinsei0916/tfjs/blob/d04e6d02360d1cf0ca4dfeab8181e92430e7d709/tfjs-backend-cpu/src/kernels/Scatter_impl.ts#L68-L71), so I think this is fine.

Reply:
[TensorFlow doc](https://www.tensorflow.org/api_docs/python/tf/raw_ops/SparseToDense) states that if `sparse_values` is scalar, the value is assigned to all sparce indices.

So as far as I understand, `sparse_indices` could take more than one index while `sparse_values` is a scalar, and the loop will in turn run more than once.